### DISCUSSION TITLE
Promote `Option` to a compiler-built-in type

### DISCUSSION THREAD

**Discussion by tranzystorekk** [gleam-lang/gleam#1334] [Ideas & suggestions]
This topic has not seen a good discussion so I've added it here.

`Option` is considered a fundamental concept in functional-oriented languages, useful in expressing many fallible operations like obtaining elements from maps, representing results of iterator operations etc.
Therefore it seems suitable that they be available in the Gleam prelude, i.e. as a core language type without the need to import them.

The proposal is as follows:

- Make the compiler insert inline `Option` usage/definitions wherever the type is used, similar to `Result`
- Remove the `Option` definition from `gleam/option`
- Add `Option` definitions/code to the js prelude
- (Lower priority) Rewrite stdlib APIs that use `Result(a, Nil)` to `Option`

Pros:

- Simplified APIs in various places
- `Option` is more expressive than `Result(a, Nil)` in most situations
- Less hassle with importing `gleam/option`, e.g. in stdlib itself

Cons:

- We currently only have `try` expression support for `Result`, so some code would have to be rewritten with `option.to_result(Nil)`

**Comment by TanklesXL** [comment]
Seconded! This would also unblock what I wanted to do for 
https://github.com/gleam-lang/stdlib/issues/214

Only thing I’m not sure I agree on is the replacing usage of `Result(a,Nil)` but I have no strong opinions on the matter

**Comment by tranzystorekk** [comment]
A reference implementation I've been working on can be inspected in this branch: https://github.com/tranzystorek-io/gleam/tree/option-as-builtin
